# WeIntern AIML Internship — Week 1
## Task 1: Student Performance Dataset Analysis

| | |
|---|---|
| **Intern** | Vratika Kumawat |
| **Organization** | WeIntern Pvt Ltd |
| **Domain** | AI/ML |
| **Tools Used** | Python, NumPy, Pandas, Matplotlib, Seaborn |
| **Platform** | Google Colab |
| **Dataset** | Student Performance Dataset (Synthetic) |

## Dataset Summary
- 205 rows, 12 columns.
- Contains numerical, categorical and identifier columns
- Target variable: FinalScore
- Dataset had missing values, duplicates and inconsistent labels
which were all handled during cleaning.
- Source: Synthetically generated dataset created for
  internship practice purposes using Python's random module.
  It simulates realistic student performance patterns
  including intentional missing values, duplicates,
  and inconsistent labels for cleaning practice.

## Step 1 — Import Libraries

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Step 2 — Load and preview the Dataset

In [3]:
df = pd.read_csv("student_performance.csv")

print("\nFirst 5 rows:")
df.head()


First 5 rows:


,StudentID,Name,Gender,Age,AttendancePct,StudyHoursPerWeek,AssignmentScore,MidtermScore,FinalScore,ParentalEducation,InternetAccess,ExtraActivities
0,196,Ananya,F,18,54.3,1.9,30.0,25.0,20.0,No Formal Education,NaN,Yes
1,56,Arjun,M,19,98.7,15.4,60.6,49.3,58.8,High School,No,No
2,79,Karan,Male,17,55.4,4.3,30.0,25.0,20.0,Bachelor's,YES,Yes
3,121,Ananya,female,20,83.4,1.2,30.0,25.0,20.0,Master's,YES,No
4,149,Vivek,male,21,NaN,8.2,33.9,40.3,37.9,PhD,yes,Yes


In [4]:
print("Last 5 rows:")
df.tail()

Last 5 rows:


,StudentID,Name,Gender,Age,AttendancePct,StudyHoursPerWeek,AssignmentScore,MidtermScore,FinalScore,ParentalEducation,InternetAccess,ExtraActivities
200,108,Swati,female,16,90.8,9.2,45.0,44.7,46.9,Master's,YES,No
201,18,Dev,male,17,66.3,11.3,40.7,47.1,36.9,Master's,yes,Yes
202,134,Neha,female,18,58.8,1.9,30.0,25.0,20.0,Bachelor's,YES,No
203,151,Siddharth,male,19,67.8,13.9,51.0,47.9,55.0,No Formal Education,Yes,No
204,160,Harish,Male,21,61.0,14.5,48.9,50.6,46.9,PhD,Yes,No


## Step 3 — Check Shape
### Check dataset shape using number of rows and columns.

In [5]:
print(f"Dataset shape: {df.shape[0]} rows × {df.shape[1]} columns")

Dataset shape: 205 rows × 12 columns


## Step 4 — Check Info
### Inspect data types using df.info().

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   StudentID          205 non-null    int64  
 1   Name               205 non-null    object 
 2   Gender             205 non-null    object 
 3   Age                205 non-null    int64  
 4   AttendancePct      194 non-null    float64
 5   StudyHoursPerWeek  193 non-null    float64
 6   AssignmentScore    205 non-null    float64
 7   MidtermScore       205 non-null    float64
 8   FinalScore         205 non-null    float64
 9   ParentalEducation  195 non-null    object 
 10  InternetAccess     200 non-null    object 
 11  ExtraActivities    205 non-null    object 
dtypes: float64(5), int64(2), object(5)
memory usage: 19.3+ KB


#### The dataset contains a mix of numerical and categorical features. Some columns have missing values. Therefore, data cleaning and preprocessing are necessary before analysis.

## Step 5 — Feature Understanding Table

| Column | Type | Role | Notes |
|--------|------|------|-------|
| StudentID | Identifier | ID only | Not useful for analysis |
| Name | Identifier | ID only | Not useful for analysis |
| Gender | Categorical | Demographic | object dtype |
| Age | Numerical | Demographic | int64, range 16-22 |
| AttendancePct | Numerical | Input Feature | float64, has 11 missing values |
| StudyHoursPerWeek | Numerical | Input Feature | float64, has 12 missing values |
| AssignmentScore | Numerical | Performance | float64, score 0-100 |
| MidtermScore | Numerical | Performance | float64, score 0-100 |
| FinalScore | Numerical | TARGET variable | float64, main outcome |
| ParentalEducation | Categorical | Background | object, has 10 missing values |
| InternetAccess | Categorical | Resource | object, has 5 missing values |
| ExtraActivities | Categorical | Engagement | object, Yes/No |

## Step 6 — Check Missing Values.

In [7]:
missing = df.isnull().sum()
missing_pct = (missing / len(df) * 100).round(2)

missing_report = pd.DataFrame({
    "Missing Count": missing,
    "Missing %": missing_pct
})

print(missing_report[missing_report["Missing Count"] > 0])

                   Missing Count  Missing %
AttendancePct                 11       5.37
StudyHoursPerWeek             12       5.85
ParentalEducation             10       4.88
InternetAccess                 5       2.44


In [8]:
# Total missing values in entire dataset
print(f"Total missing values: {missing.sum()}")

Total missing values: 38


#### The dataset contains missing values in selected numerical and categorical variables, while several columns are complete. Appropriate preprocessing and imputation strategies will be applied during data cleaning to prepare the dataset for accurate analysis.

## Step 7 — Check and eliminate duplicate rows

In [9]:
# Count duplicate rows
duplicates = df.duplicated().sum()
print(f"Number of duplicate rows: {duplicates}")

Number of duplicate rows: 5


In [10]:
# Remove duplicate rows
df = df.drop_duplicates()

print(f"Rows before: 205")
print(f"Rows after : {df.shape[0]}")

Rows before: 205
Rows after : 200


## Step 8 - Check Inconsistent Labels.


In [11]:
# Get all categorical columns
cat_cols = df.select_dtypes(include="object").columns

# Loop through each one and print unique values
for col in cat_cols:
    print(f"{col}: {df[col].unique()}")
    print()

Name: ['Ananya' 'Arjun' 'Karan' 'Vivek' 'Vikram' 'Divya' 'Manish' 'Neha' 'Meera'
 'Raj' 'Dev' 'Nikhil' 'Harish' 'Rahul' 'Rohan' 'Piyush' 'Riya' 'Ravi'
 'Swati' 'Simran' 'Aarav' 'Amit' 'Kavya' 'Pooja' 'Tanvi' 'Komal' 'Sakshi'
 'Nisha' 'Sneha' 'Preeti' 'Naveen' 'Aditi' 'Isha' 'Anjali' 'Suresh'
 'Siddharth' 'Shreya' 'Arun' 'Deepak' 'Priya']

Gender: ['F' 'M' 'Male' 'female' 'male' 'Female']

ParentalEducation: ['No Formal Education' 'High School' "Bachelor's" "Master's" 'PhD' nan]

InternetAccess: [nan 'No' 'YES' 'yes' 'Yes']

ExtraActivities: ['Yes' 'No']



# Data Cleaning

## Step 9 — Standardize Inconsistent Labels.

In [12]:
# Step 1 — make everything lowercase first (removes case differences)
df["Gender"] = df["Gender"].str.strip().str.lower()
df["InternetAccess"] = df["InternetAccess"].str.strip().str.lower()

# Step 2 — now replace everything to a standard label
df["Gender"] = df["Gender"].replace({"m": "Male", "male": "Male",
                                      "f": "Female", "female": "Female"})

df["InternetAccess"] = df["InternetAccess"].replace({"yes": "Yes", "no": "No"})

# Verification
print("Gender values after:", df["Gender"].unique())
print("InternetAccess values after:", df["InternetAccess"].unique())

Gender values after: ['Female' 'Male']
InternetAccess values after: [nan 'No' 'Yes']


## After Standardisation

In [13]:
cat_cols = df.select_dtypes(include="object").columns
for col in cat_cols:
    print(f"{col}: {df[col].unique()}")
    print()

Name: ['Ananya' 'Arjun' 'Karan' 'Vivek' 'Vikram' 'Divya' 'Manish' 'Neha' 'Meera'
 'Raj' 'Dev' 'Nikhil' 'Harish' 'Rahul' 'Rohan' 'Piyush' 'Riya' 'Ravi'
 'Swati' 'Simran' 'Aarav' 'Amit' 'Kavya' 'Pooja' 'Tanvi' 'Komal' 'Sakshi'
 'Nisha' 'Sneha' 'Preeti' 'Naveen' 'Aditi' 'Isha' 'Anjali' 'Suresh'
 'Siddharth' 'Shreya' 'Arun' 'Deepak' 'Priya']

Gender: ['Female' 'Male']

ParentalEducation: ['No Formal Education' 'High School' "Bachelor's" "Master's" 'PhD' nan]

InternetAccess: [nan 'No' 'Yes']

ExtraActivities: ['Yes' 'No']



## Step 10 — Fill Missing Values

In [14]:
# Numerical columns → fill with median
for col in ["StudyHoursPerWeek", "AttendancePct"]:
    median_val = df[col].median()
    df[col] = df[col].fillna(median_val)
    print(f"Filled '{col}' with median: {median_val:.2f}")

# Categorical columns → fill with 'Unknown'
for col in ["ParentalEducation", "InternetAccess"]:
    df[col] = df[col].fillna("Unknown")
    print(f"Filled '{col}' with 'Unknown'")

Filled 'StudyHoursPerWeek' with median: 10.90
Filled 'AttendancePct' with median: 69.90
Filled 'ParentalEducation' with 'Unknown'
Filled 'InternetAccess' with 'Unknown'


### verify that all missing values are actually gone

In [15]:
print(df.isnull().sum())

StudentID            0
Name                 0
Gender               0
Age                  0
AttendancePct        0
StudyHoursPerWeek    0
AssignmentScore      0
MidtermScore         0
FinalScore           0
ParentalEducation    0
InternetAccess       0
ExtraActivities      0
dtype: int64


## Step 11 — Rename columns

In [16]:
# Check current column names
print(df.columns.tolist())

['StudentID', 'Name', 'Gender', 'Age', 'AttendancePct', 'StudyHoursPerWeek', 'AssignmentScore', 'MidtermScore', 'FinalScore', 'ParentalEducation', 'InternetAccess', 'ExtraActivities']


### All column names are clear and descriptive. No renaming required.

## Step 12 — Convert Data Types

In [17]:
# Check current data types
print(df.dtypes)

StudentID              int64
Name                  object
Gender                object
Age                    int64
AttendancePct        float64
StudyHoursPerWeek    float64
AssignmentScore      float64
MidtermScore         float64
FinalScore           float64
ParentalEducation     object
InternetAccess        object
ExtraActivities       object
dtype: object


### All columns have appropriate data types. No conversion required.

## Step 13 — Save the Cleaned Dataset

In [18]:
df.to_csv("student_performance_cleaned.csv", index=False)
print("Cleaned dataset saved!")

Cleaned dataset saved!


In [22]:
# Reload cleaned dataset to verify it saved correctly
clean_df = pd.read_csv("student_performance_cleaned.csv")
print("Verified! Shape:", clean_df.shape)
clean_df.head()

Verified! Shape: (200, 12)


,StudentID,Name,Gender,Age,AttendancePct,StudyHoursPerWeek,AssignmentScore,MidtermScore,FinalScore,ParentalEducation,InternetAccess,ExtraActivities
0,196,Ananya,Female,18,54.3,1.9,30.0,25.0,20.0,No Formal Education,Unknown,Yes
1,56,Arjun,Male,19,98.7,15.4,60.6,49.3,58.8,High School,No,No
2,79,Karan,Male,17,55.4,4.3,30.0,25.0,20.0,Bachelor's,Yes,Yes
3,121,Ananya,Female,20,83.4,1.2,30.0,25.0,20.0,Master's,Yes,No
4,149,Vivek,Male,21,69.9,8.2,33.9,40.3,37.9,PhD,Yes,Yes


## Assumptions Made During Cleaning

1. Missing numerical values (AttendancePct, StudyHoursPerWeek)
   were filled with median — assumed median best represents
   a typical student without being affected by outliers.

2. Missing categorical values (ParentalEducation, InternetAccess)
   were filled with 'Unknown' — assumed it is better to honestly
   label missing data than guess a category.

3. Duplicate rows were removed keeping the first occurrence —
   assumed they were data entry errors, not real students.

4. Inconsistent gender labels (M, male, Male) were standardized
   to Male/Female — assumed they all refer to the same categories.

5. Original dataset was kept unchanged and a separate cleaned
   version was saved — assumed raw data should always be preserved.

6. Column names were assumed to be clear and descriptive
   enough — no renaming was required.

7. All column data types were assumed to be correct
   (int64, float64, object) — no type conversion was required.

## Step 14 - Descriptive Statistics

In [19]:
df.describe().round(2)

,StudentID,Age,AttendancePct,StudyHoursPerWeek,AssignmentScore,MidtermScore,FinalScore
count,200.00,200.00,200.00,200.00,200.00,200.00,200.00
mean,100.50,19.10,70.09,10.32,38.88,37.62,36.74
std,57.88,2.03,16.37,5.45,9.18,10.38,10.86
min,1.00,16.00,40.40,1.10,30.00,25.00,20.00
25%,50.75,17.00,56.60,5.40,30.00,27.15,28.08
50%,100.50,19.00,69.90,10.90,35.90,37.00,36.95
75%,150.25,21.00,82.88,14.58,45.02,44.90,44.82
max,200.00,22.00,99.80,19.70,67.20,63.20,66.10


### NumPy Analysis of Final Scores

In [20]:
scores_np = np.array(df["FinalScore"])
print(f"NumPy Mean  : {np.mean(scores_np):.2f}")
print(f"NumPy Std   : {np.std(scores_np):.2f}")
print(f"NumPy Min   : {np.min(scores_np):.2f}")
print(f"NumPy Max   : {np.max(scores_np):.2f}")

NumPy Mean  : 36.74
NumPy Std   : 10.84
NumPy Min   : 20.00
NumPy Max   : 66.10


## Step 15 — Group Comparisons
### Analysing how different groups perform on FinalScore

In [21]:
print("Average Final Score by Internet Access:")
print(df.groupby("InternetAccess")["FinalScore"].mean().round(2).to_frame())
print()
print("Average Final Score by Extra Activities:")
print(df.groupby("ExtraActivities")["FinalScore"].mean().round(2).to_frame())
print()
print("Average Final Score by Gender:")
print(df.groupby("Gender")["FinalScore"].mean().round(2).to_frame())

Average Final Score by Internet Access:
                FinalScore
InternetAccess            
No                   36.35
Unknown              41.54
Yes                  36.68

Average Final Score by Extra Activities:
                 FinalScore
ExtraActivities            
No                    36.78
Yes                   36.71

Average Final Score by Gender:
        FinalScore
Gender            
Female       36.25
Male         37.24


## Observations

1. The average final score is 36.74 out of 100, suggesting overall
   academic performance is below average across the dataset.

2. Attendance ranges from 40% to 99.8%, showing a wide gap between
   the most and least attending students.

3. Study hours range from 1.1 to 19.7 hours per week, indicating
   some students study significantly more than others.

4. The median final score (36.95) is very close to the mean (36.74),
   suggesting the score distribution is fairly balanced without
   extreme outliers pulling it in either direction.

5. Assignment scores start at a minimum of 30, while final scores
   go as low as 20, suggesting some students perform worse in
   finals than in assignments.

6. 25% of students attend less than 56.6% of classes, which is
   quite low and may negatively impact their final scores.

7. Study hours average 10.32 per week — students in the top 25%
   study 14.58+ hours, nearly 3x more than the bottom 25% at 5.4 hours.

8. Students with internet access (36.68) scored slightly higher
   than those without (36.35), though the difference is small
   and may not be significant.

9. Extra activities show almost no difference in final scores
   (Yes: 36.71 vs No: 36.78), suggesting participation in
   extra activities neither helps nor hurts academic performance
   in this dataset.

10. Students with 'Unknown' internet access status scored highest
    on average (41.54), however this group is very small (missing
    values filled as Unknown), so this result is likely unreliable
    and should not be interpreted as a real pattern.

## Cleaned Dataset Successfully Saved

After completing all data cleaning and preprocessing steps, the cleaned dataset was exported and saved as a separate CSV file. Maintaining a dedicated cleaned dataset preserves the integrity of the original raw data while providing a reliable and analysis-ready version for further exploration, visualization, and future machine learning tasks.

This approach follows best practices in data analytics by ensuring reproducibility, transparency, and clear separation between raw and processed data.